pandas/core/frame.py

```diff
@@ -4151,10 +4146,28 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
                 # tuples are always considered keys, never as list-likes
                 if col not in self:
                     missing.append(col)
-            elif (not isinstance(col, (ABCIndexClass, ABCSeries,
-                                       np.ndarray, list))
-                  or getattr(col, 'ndim', 1) > 1):
-                raise ValueError(err_msg)
+            elif isinstance(col, (ABCIndexClass, ABCSeries,
+                                  np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif is_list_like(col, allow_sets=False):
+                # various iterators/generators are hashable, but should not
+                # raise a KeyError
+                tipo = type(col)
+                raise ValueError(err_msg + ' Received column of '
+                                 'type {}'.format(tipo))
```

Comment:
I added this extra branch to give a more sensible error for the iterator/generator cases. It's maybe worth noting that it would be easy to re-add the capability to consume list-likes here, because tuples now *always* enter the first branch.

Reply:
I am not fully sure we should use `is_list_like` here either. You can have custom objects that are hashable, but also iterable (like a tuple is also the combination). And once it is iterable, `is_list_like` will give True.

Reply:
And the line we draw is on hashable (at least, if it is not hashable, the indexing machinery simply doesn't work), not on iterable (tuples are iterable, strings are iterable)

I am not saying I would use it myself, but there are certainly reasonable examples of things you can put in an object dtype array that are iterable. One example is a shapely geometry object which are iterable (iterate through coordinates points; strictly spoken, they are currently not hashable, but that is something they plan to fix).

Given 0.24.1, I think we might want to release today. Are you fine with us adding some small changes here to get us to merge it during the day? (keeping the broad rationale of course)

Reply:
@jorisvandenbossche 
Feel free to make any desired changes to this PR, my strong preference is no `KeyError` for iterables though.

> there are certainly reasonable examples of things you can put in an object dtype array that are iterable

Why should an object dtype array (which is mutable) be hashable?

My suggestion would be that you try to come up with a hashable/iterable example that would also be list-like (from the POV of `is_list_like`, which would likely exclude various custom types, unless they directly inherit from `abc.Iterable`), before you start changing things around, but feel free.

Reply:
> Why should an object dtype array (which is mutable) be hashable?

Sorry, I just generally meant object dtype array-like, like an Index or Series. So that it can make sense to put hashable but iterable objects in an Index or Series

Reply:
> My suggestion would be that you try to come up with a hashable/iterable example that would also be list-like

As I said, eg a shapely geometry. 
(anything that is iterable is regarded 'list-like' in the eyes of `is_list_like`)

Reply:
OK, didn't see the shapely example. Maybe use something like `is_list_like(col) and not hasattr('__hash__')` for the last `elif`?

Reply:
Didn't see your comments pop up, and edited my answer independently about the shaply case

In any case, **I can't work on this any more today**. Failures seem to be due to dict-views being hashable on `PY2`. That case could simply be removed from the parametrization, it was an extra mile I tried to go.